File: java/spring-gcp/api-logging-gcp/src/main/java/azoth/gcp/logging/controller/ProductController.java
package azoth.gcp.logging.controller;

import azoth.gcp.logging.service.ProductService;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

@RestController
@RequestMapping(path = "/products", produces = MediaType.APPLICATION_JSON_VALUE)
public class ProductController {
    ProductService service;

    public ProductController(ProductService service) {
        this.service = service;
    }

    @GetMapping("/{id}")
    public ProductService.Product findById(@PathVariable("id") int id) {
        return service.findById(id);
    }

    @GetMapping("")
    public List<ProductService.Product> findAll() {
        return service.findAll();
    }
}

File: kotlin-spring/reactive-api/src/main/kotlin/azoth/reactive/controller/rest/ProductController.kt
package azoth.reactive.controller.rest

import azoth.reactive.model.ProductDTO
import azoth.reactive.model.dto
import azoth.reactive.model.entity
import azoth.reactive.repo.ProductRepository
import kotlinx.coroutines.reactor.awaitSingleOrNull
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.web.server.ResponseStatusException

@RestController
@RequestMapping(path = ["/products"])
class ProductController (@Autowired val repo: ProductRepository){

    @PostMapping(path = [""])
    suspend fun create(@RequestBody dto: ProductDTO): ResponseEntity<Any> {
       repo.save(dto.entity()).awaitSingleOrNull()

        return ResponseEntity.status(HttpStatus.CREATED).build()
    }

    @GetMapping(path = ["/{id}"])
    suspend fun getById(@PathVariable id: String): ProductDTO {
        val result = repo.findById(id).awaitSingleOrNull()
            ?: throw ResponseStatusException(HttpStatus.NOT_FOUND, "Product with $id couldn't be found")

        return result.dto()
    }
}